src/transform/src/typecheck.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Check that the visible type of each query has not been changed
+
+use std::{cell::RefCell, collections::BTreeMap};
+
+use mz_compute_client::types::dataflows::BuildDesc;
+use mz_expr::{Id, OptimizedMirRelationExpr};
+use mz_repr::ColumnType;
+use tracing::{error, info, warn};
+
+use crate::TransformError;
+
+/// Check that the visible type of each query has not been changed
+#[derive(Debug)]
+pub struct Typecheck {
+    /// The known types of the queries so far
+    pub ctx: RefCell<BTreeMap<Id, Vec<ColumnType>>>,
+}
+
+impl Default for Typecheck {
+    fn default() -> Self {
+        Self {
+            ctx: RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+impl crate::Transform for Typecheck {
+    #[tracing::instrument(
+        target = "optimizer"
+        level = "trace",
+        skip_all,
+        fields(path.segment = "typecheck")
+    )]
+    fn transform(
+        &self,
+        relation: &mut mz_expr::MirRelationExpr,
+        _args: crate::TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        let ctx = self.ctx.borrow();
+
+        if let Err(err) = relation.typecheck(&ctx) {
+            warn!("TYPE ERROR: {}\nIN UNKNOWN QUERY:\n{:#?}", err, relation);
+        }
```

Comment:
You can add a variant to `TransformError` and propagate the error this way, or use the `Internal` variant. Type errors are not something the user can fix, so categorizing it as internal errors should be fine.

Reply:
Switching to `TransformError::Internal`.